scipy/integrate/_tanhsinh.py

```diff
@@ -321,53 +327,54 @@ def _tanhsinh(f, a, b, *, args=(), log=False, maxfun=None, maxlevel=None,
+        c = xp.reshape((xp_ravel(a) + xp_ravel(b))/2, a.shape)
+        inf_a, inf_b = xp.isinf(a), xp.isinf(b)
+        c[inf_a] = b[inf_a] - 1.  # takes care of infinite a
+        c[inf_b] = a[inf_b] + 1.  # takes care of infinite b
+        c[inf_a & inf_b] = 0.  # takes care of infinite a and b
         temp = eim._initialize(f, (c,), args, complex_ok=True,
-                               preserve_shape=preserve_shape)
+                               preserve_shape=preserve_shape, xp=xp)
     f, xs, fs, args, shape, dtype, xp = temp
-    a = np.broadcast_to(a, shape).astype(dtype).ravel()
-    b = np.broadcast_to(b, shape).astype(dtype).ravel()
+    a = xp_ravel(xp.astype(xp.broadcast_to(a, shape), dtype))
+    b = xp_ravel(xp.astype(xp.broadcast_to(b, shape), dtype))
 
     # Transform improper integrals
-    a, b, a0, negative, abinf, ainf, binf = _transform_integrals(a, b)
+    a, b, a0, negative, abinf, ainf, binf = _transform_integrals(a, b, xp)
 
     # Define variables we'll need
     nit, nfev = 0, 1  # one function evaluation performed above
-    zero = -np.inf if log else 0
-    pi = dtype.type(np.pi)
+    zero = -xp.inf if log else 0
+    pi = xp.asarray(xp.pi, dtype=dtype)[()]
     maxiter = maxlevel - minlevel + 1
-    eps = np.finfo(dtype).eps
+    eps = xp.finfo(dtype).eps
     if rtol is None:
-        rtol = 0.75*np.log(eps) if log else eps**0.75
+        rtol = 0.75*math.log(eps) if log else eps**0.75
 
-    Sn = np.full(shape, zero, dtype=dtype).ravel()  # latest integral estimate
-    Sn[np.isnan(a) | np.isnan(b) | np.isnan(fs[0])] = np.nan
-    Sk = np.empty_like(Sn).reshape(-1, 1)[:, 0:0]  # all integral estimates
-    aerr = np.full(shape, np.nan, dtype=dtype).ravel()  # absolute error
-    status = np.full(shape, eim._EINPROGRESS, dtype=int).ravel()
-    h0 = np.real(_get_base_step(dtype=dtype))  # base step
+    Sn = xp_ravel(xp.full(shape, zero, dtype=dtype))  # latest integral estimate
+    Sn[xp.isnan(a) | xp.isnan(b) | xp.isnan(fs[0])] = xp.nan
+    Sk = xp.reshape(xp.empty_like(Sn), (-1, 1))[:, 0:0]  # all integral estimates
+    aerr = xp_ravel(xp.full(shape, xp.nan, dtype=dtype))  # absolute error
+    status = xp_ravel(xp.full(shape, eim._EINPROGRESS, dtype=xp.int32))
+    h0 = _get_base_step(dtype, xp)
+    h0 = xp_real(h0) # base step
 
     # For term `d4` of error estimate ([1] Section 5), we need to keep the
     # most extreme abscissae and corresponding `fj`s, `wj`s in Euler-Maclaurin
     # sum. Here, we initialize these variables.
-    xr0 = np.full(shape, -np.inf, dtype=dtype).ravel()
-    fr0 = np.full(shape, np.nan, dtype=dtype).ravel()
-    wr0 = np.zeros(shape, dtype=dtype).ravel()
-    xl0 = np.full(shape, np.inf, dtype=dtype).ravel()
-    fl0 = np.full(shape, np.nan, dtype=dtype).ravel()
-    wl0 = np.zeros(shape, dtype=dtype).ravel()
-    d4 = np.zeros(shape, dtype=dtype).ravel()
+    xr0 = xp_ravel(xp.full(shape, -xp.inf, dtype=dtype))
```

Comment:
It's not bad; it makes the `eim` framework easier to work because the arrays having a standard number of dimensions (the 0th dimension is the batch dimension and the 1st dimension is an additional dimension that may be used by the algorithm) rather than whatever dimensionality the user passed in.